File: codejam/p2018_lucky_dip.cpp
#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <vector>
#include <functional>
#include <algorithm>

using namespace std;

double geo_seq_sum(double a, double r, int n) {
	return (a * (1 - pow(r, n))) / (1 - r);
}	


double work() {
	int n, redip;
	scanf("%d", &n);
	scanf("%d", &redip);

	int dip = redip + 1;
	vector<int> num_list;
	for (int i = 0; i < n; i++) {
		int item;
		scanf("%d", &item);
		num_list.push_back(item);
	}
	num_list.resize(n);
	sort(num_list.begin(), num_list.end(), greater<int>());

	double expected_value = 0;

	int idx = 0;
	int cur, cur_count;
	cur = num_list[0];
	cur_count = 0;
	while (1) {
		if (idx == n) {
			return cur;
		}

		printf("idx: %d, n: %d, cur: %d, num: %d\n", idx, n, cur, num_list[idx]);
		if (cur > num_list[idx]) {
			double prob = (double)cur_count / (double)n;
			double pick_other_prob = (double)(n-idx) / (double)n;
			double g_sum = geo_seq_sum(prob, pick_other_prob, dip);
			expected_value += cur * g_sum;
			printf("cur: %d, g_sum: %f\n", cur, g_sum);
			break;
		}
		else {
			cur_count++;
			idx++;
		}
	}

	cur = num_list[idx];
	cur_count = 0;
	for (int i = idx; i < n+1; i++) {
		if (i == n) {
			double p1 = pow((double)cur_count / (double)n, dip);
			printf("cur: %d, p1: %f\n", cur, p1);
			expected_value += cur * p1;
		}
		else if (cur > num_list[i]) {
			double p1 = pow((double)(n-(i-cur_count)) / (double)n, dip);
			double p2 = pow((double)(n-i) / (double)n, dip);

			printf("cur: %d, p1-p2: %f\n", cur, p1-p2);	
			expected_value += cur * (p1 - p2);
			cur = num_list[i];
			cur_count = 1;
		}
		else {
			cur_count++;
		}
	}

	return expected_value;
}

int main() {
	int test_case;
	scanf("%d\n", &test_case);
	for (int i = 0; i < test_case; i++) {
		printf("Case #%d: %f\n", i+1, work());
	}

	return 0;
}

File: codejam/p2018_lucky_dip.py
import sys
import math


test_case = int(sys.stdin.readline())

def geometric_seq_sum(a, r, n):
	return (a*(1 - math.pow(r, n))) / (1 - r)

def work():
	n, redip = map(int, sys.stdin.readline().rstrip().split())
	dip = redip + 1

	num_list = list(map(int, sys.stdin.readline().rstrip().split()))
	num_list.sort(reverse=True)	# O(nlogn)

	expected_value = 0

	v = 0

	# pick max value
	idx = 0
	cur = num_list[0]	
	cur_count = 0
	while True:
		if idx == n:
			return cur	
		if cur > num_list[idx]:
			prob = cur_count / n
			pick_other_prob = (n-idx) / n	
			g_sum = geometric_seq_sum(prob, pick_other_prob, dip)
			v += g_sum
			expected_value += cur * g_sum
			# print(expected_value)
			break
		else:
			cur_count += 1
			idx += 1

	cur = num_list[idx]
	cur_count = 0
	for i in range(idx, n+1):
		if i == n:
			p1 = math.pow((cur_count / n), dip)
			v += p1
			# print(cur, cur_count, n, dip)
			expected_value += cur * p1
			# print(expected_value)
		elif cur > num_list[i]:
			p1 = math.pow(((n-(i-cur_count)) / n), dip)
			p2 = math.pow(((n-i) / n), dip)
			# print(cur, n-(i-cur_count), n, n-i, n, dip)
			expected_value += cur * (p1 - p2)
			
			# print(expected_value)
			v += p1-p2
			cur = num_list[i]
			cur_count = 1
		else:
			cur_count += 1
	
	# print(v)
	return expected_value

for i in range(test_case):
	sys.stdout.write('Case #%d: %.9f\n' % (i+1, work()))
